### ISSUE TITLE
Review Lab2

### ISSUE THREAD

**Issue by LeonardoTredese** [saccuz/Computational_Intelligence#4]
Major
====
- [You are not storing the fittest individual that you have ever found](https://github.com/saccuz/Computational_Intelligence/blob/64919ded2eb7ceea9502c2e123aa5c93f54fcbf2/lab2/search.py#L90).  Even if keep the best solution of the population, you might mutate the best solution in a worse one and loose it in the final result.
- [You are not truly performing Elitims](https://github.com/saccuz/Computational_Intelligence/blob/64919ded2eb7ceea9502c2e123aa5c93f54fcbf2/lab2/search.py#L92) as the best element still has the chance to be mutated.
- [You completely overturn the population with offspring](https://github.com/saccuz/Computational_Intelligence/blob/64919ded2eb7ceea9502c2e123aa5c93f54fcbf2/lab2/search.py#L91). This mean that even if you have set [population size to 30](https://github.com/saccuz/Computational_Intelligence/blob/64919ded2eb7ceea9502c2e123aa5c93f54fcbf2/lab2/lab2.py#L16) the population will always be 20 after the first iteration. In strategy one in the new population you will throw away worst individuals and so you should compare the current population with the offspring as well.

Minor
====
- Please insert your depenencies in the [pyproject.toml](https://github.com/saccuz/Computational_Intelligence/blob/64919ded2eb7ceea9502c2e123aa5c93f54fcbf2/pyproject.toml) so that is easier to install your dependencies
-   When computing the fitness you are [dereferencing the space list](https://github.com/saccuz/Computational_Intelligence/blob/64919ded2eb7ceea9502c2e123aa5c93f54fcbf2/lab2/search.py#L29) using an index, this is O(n) in python because lists are composed of linked nodes, they are not like arrays. To iterate 2 lists at the same time you can use the  [zip function](https://docs.python.org/3.10/library/functions.html#zip).

Suggestions
==========
- You can easily and quickly [count collisions](https://github.com/saccuz/Computational_Intelligence/blob/64919ded2eb7ceea9502c2e123aa5c93f54fcbf2/lab2/search.py#L14) using the [Counter object](https://docs.python.org/3.10/library/collections.html#collections.Counter) in python. More over this object removes the necessity to count the number of different individuals in the [fitness function](https://github.com/saccuz/Computational_Intelligence/blob/64919ded2eb7ceea9502c2e123aa5c93f54fcbf2/lab2/search.py#L35).
-  If you use list comprehensions in a for loop python would first generate a whole list and then iterate over it. You do this when [initializing with the population](https://github.com/saccuz/Computational_Intelligence/blob/64919ded2eb7ceea9502c2e123aa5c93f54fcbf2/lab2/search.py#L69). Since you do this only once in a run this will not have a terrible impact on execution time. Anyway to avoid creating a full list you can use a generator by simply changing [... for x in ...] to (... for x in ...), or even create the population directly with a list comprehension which is optimized in python.

Comments
========
The representation is good and correct, although it might not scale due to memory footprint for larger spaces.
Your mutation slightly changes an individual, which is fine as it would not overturn a good solution. The single cut crossover is ok.

I like that you start from almost empty and incomplete solutions, using the (# elem, - # collisions) is concretely useful at least in initial iterations.

**Comment by saccuz** [comment]
I do not agree with the first point of "major". In fact taking the fittest individual of the population and re-adding it inside the new population (after I throw away the old generation) make the fittest of the old generation survive. And if no fitter individual rise, I will have the saved one as the fittest once again.
The second and the third point of "major" are done on purpose, in order to increment the randomness of the evolution to escape the plateau we reach pretty earlier in the execution.

I really thank you for the other suggestions you have provided!

**Comment by LeonardoTredese** [comment]
You are right about the first point, I made the mistake while reading the code. I took a closer look and  the best solution is not subject to any kind of mutation, so the best is not lost unless a better individual is introduced. Moreover this disrupts point 2 of 'major' as well. By the way since you discard the whole population you can avoid to sort it and just get the fittest element with `max`.

Its still is not clear to me why you specify a population of  POPULATION_SIZE for a single iteration and then change it to  OFFSPRING_SIZE.  I ran the search 3 times and measured the average number of collisions for different problem sizes to see if this created and improvement. Here are the results.

### POPULATION_SIZE = 30, OFFSPRING_SIZE = 20
| N = 5 | N = 10 |  N = 50 | N = 100 | N = 500 | N = 1000 |
|---------|-----------|-----------|------------|------------|---------------|
| 0.0 |0.33|39.66|126.66|1279.8| 3297.8 |

### POPULATION_SIZE = 20, OFFSPRING_SIZE = 20
| N = 5 | N = 10 |  N = 50 | N = 100 | N = 500 | N = 1000 |
|---------|-----------|-----------|------------|------------|---------------|
| 0.0 | 0.33 | 42.33 | 125.66 | 1302.33 | 3235.66|

If I correctly understood, you would like a bigger population at the beginning to favor exploration and then a smaller one to favor exploitation. The results showed that doing so only in the initial iteration does not differ much than not doing it at all. You could try to reduce the population size by adaptation or after a number of generations grater than 1.

If you wish I can share the code I used to create the tables .
